src/Microsoft.AspNet.OData.Shared/ODataAPIHandler.cs

```diff
@@ -64,129 +69,245 @@ public abstract class ODataAPIHandler<TStructuralType>: IODataAPIHandler where T
                     {
-                        foreach (object item in lst)
+                        if (item.GetType().IsPrimitive)
                         {
-                            if (item.GetType().IsPrimitive)
-                            {
-                                break;
-                            }
-
-                            CheckAndApplyODataId(item, model);
+                            break;
                         }
-                    }
-                    else
-                    {
-                        CheckAndApplyODataId(propVal, model);
+
+                        CopyObjectProperties(item, model, nestedHandler, apiHandlerFactory);
                     }
                 }
+                else
+                {
+                    CopyObjectProperties(navPropVal, model, nestedHandler, apiHandlerFactory);
+                }
             }
         }
 
-        private object ApplyODataIdOnContainer(object obj, ODataPath odataPath)
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1031:DoNotCatchGeneralExceptionTypes", Justification = "We don't have a specific exception to catch.")]
+        internal static void ApplyHandlers(IODataAPIHandler odataApiHandler, IODataAPIHandler odataIdContainerHandler, object resource, IDictionary<string, object> keys, List<string> navigationProperties, out bool failedOperation)
         {
-            KeySegment keySegment = odataPath.LastOrDefault() as KeySegment;
-            Dictionary<string, object> keys = keySegment?.Keys.ToDictionary(x => x.Key, x => x.Value);
-            if (keys != null)
+            failedOperation = false;
+            DataModificationOperationKind operation = DataModificationOperationKind.Insert;
+
+            try
             {
-                TStructuralType returnedObject;
-                string error;
-                if (this.ParentObject.Equals(obj))
+                object[] handlerParams = new object[] { keys, null, null };
+
+                IODataAPIHandler handlerForGet = odataApiHandler;
+                ODataAPIResponseStatus getResponse = ODataAPIResponseStatus.NotFound;
+                object getObject = null;
+
+                if (odataIdContainerHandler != null)
                 {
-                    if (this.TryGet(keys, out returnedObject, out error) == ODataAPIResponseStatus.Success)
+                    handlerForGet = odataIdContainerHandler;
+                    getResponse = (ODataAPIResponseStatus)handlerForGet.GetType().GetMethod(nameof(TryGet)).Invoke(handlerForGet, handlerParams);
+                    getObject = handlerParams[1];
+                }
+
+                if (getResponse == ODataAPIResponseStatus.Success)
+                {
+                    operation = DataModificationOperationKind.Link;
+
+                    CopyProperties(getObject, resource, navigationProperties);
+
+                    object[] addRelatedObjectParams = new object[] { getObject, null };
```

Comment:
```suggestion
                    object[] relatedObjectParams = new object[] { getObject, null };
```
These variable names that sound like methods are really confusing.